### ISSUE TITLE
Deprecating SSW in scikit-bio: recommendation for alignment algorithms in Python

### ISSUE THREAD

**Issue by ebolyen** [biocore/scikit-bio#1814]
Hello everyone,

We're planning on removing the SSW module from scikit-bio. This will occur sometime before the next release (which should be sometime in October). 

It has been a general source of maintenance issues and limits our ability to target Windows platforms. While there have been some very lovely contributions such as #1709 and #1797 (from @mr-c and @theAeon  respectively), we think the best approach is to remove this module.

Instead, we recommend using @jeffdaily's library [Parasail](https://github.com/jeffdaily/parasail) and corresponding [Python bindings: jeffdaily/parasail-python](https://github.com/jeffdaily/parasail-python).

The documentation for this package is quite complete, and supports a similar usage pattern as our SSW bindings:
https://github.com/jeffdaily/parasail-python#ssw-library-emulation

The only difficulty you may encounter from this transition is converting the cigar strings (from the "traceable" algorithms, into scikit-bio alignment objects). We would be happy to accept pull requests to this end, but we don't have anything in particular planned.

If you find @jeffdaily's Parasail useful, please remember [to cite it](https://github.com/jeffdaily/parasail-python#citing-parasail) :)

**Comment by ebolyen** [comment]
And here is a worked example comparing skbio's SSW with Parasail:
https://gist.github.com/ebolyen/f759644a67d97a808f0ddb0933bcb4a0

**Comment by theAeon** [comment]
Would [this](https://github.com/theAeon/scikit-bio-parasail/) skbio-ified parasail package (that I shamelessly copied and packaged from @jvkersch) be of any use?

**Comment by ebolyen** [comment]
That is super lovely @theAeon! Thanks for sharing!

**Comment by ebolyen** [comment]
I will mention we are [looking for maintainers for this project](https://github.com/biocore/scikit-bio/pull/1809/files), so if anyone feels particularly strongly about alignment and how that should work in skbio (or other parts of scikit-bio), feel free to get in touch!

**Comment by gregcaporaso** [comment]
[pywfa](https://github.com/kcleal/pywfa) was pointed out in #254 as another performant Python-accessible aligner.

**Comment by gregcaporaso** [comment]
We've decided to delay this removal, most likely for one more release. We identified two tools that are using ``: [IAB](https://github.com/applied-bioinformatics/iab2/blob/56b597c972dcf0010987ab2affd5f9329b497f13/book/database-searching.md#defining-the-problem) and [q2-feature-classifier](https://github.com/qiime2/q2-feature-classifier/blob/ec0436f1f8fa5710f48b74048e93e765e5485e49/q2_feature_classifier/_cutter.py#L51). We suspect therefore that there may be others, so we're going to save this for 0.6.0.

**Comment by theAeon** [comment]
Also makes sense from a semver perspective for sure.

**Comment by jvkersch** [comment]
@wasimsandhu There is some discussion about this on the Parasail GitHub page, see e.g. https://github.com/jeffdaily/parasail-python/issues/72.

**Comment by qiyunzhu** [comment]
Hi all in this thread: we added a new `pair_align` function to scikit-bio to replace the old SSW wrapper and pure Python code. It is as fast as parasail without vectorization (benchmarks here: #2226), and is quite versatile too.

https://scikit.bio/docs/dev/generated/skbio.alignment.pair_align.html

To @gregcaporaso @ebolyen : The new function can be a drop-in replacement of the [SSW call in q2-feature-classifier](https://github.com/qiime2/q2-feature-classifier/blob/524f754f9b88fc741aaac111635c3452ccfc3fb5/q2_feature_classifier/_cutter.py#L40). It handles degenerate codes and it is exact (not heuristic). In my tests SSW is still faster but the results are usually sub-optimal.

**Comment by qiyunzhu** [comment]
Also, scikit-bio can now convert CIGAR strings into alignments, therefore enabling parsing the results of various external aligners, including Parasail. For example:

```python
>>> seq1 = 'GATCGTC'
>>> seq2 = 'ATCGCTC'

>>> import parasail
>>> res = parasail.nw_trace(seq1, seq2, 10, 1, parasail.nuc44)
>>> res.score
10

>>> cigar = res.cigar.decode
>>> cigar
b'1I4=1D2='

>>> from skbio.sequence import DNA
>>> from skbio.alignment import PairAlignPath, TabularMSA
>>> path = PairAlignPath.from_cigar(cigar)
>>> path
<PairAlignPath, positions: 8, segments: 4, CIGAR: '1I4M1D2M'>

>>> msa = TabularMSA.from_path_seqs(path, [DNA(seq1), DNA(seq2)])
>>> msa
TabularMSA[DNA]
---------------------
Stats:
    sequence count: 2
    position count: 8
---------------------
-GATCGTC
ATCGC-TC
```

**Comment by qiyunzhu** [comment]
Evan, you are welcome! Let us know if we can further assist with the q2-feature-classifier issue.

**Comment by gregcaporaso** [comment]
Yeah, this is great - so excited that we can finally move on from SSW. I'll get an IAB update on my list to use the new API as SSW was used there too.
